### PR TITLE
Nebula development

### PR DESCRIPTION
**Pull request by misternebula**

### CODE REVIEW

File: QSB/Finder.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+using QSB.Events;
+using QSB.Utility;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace QSB
+{
+    public static class Finder
```
Comment (Raicuparta) [inline QSB/Finder.cs:12]: From the name I'd assume this class would be just like Outer Wilds'  `Locator`, but it doesn't seem to be that at all. Everything here is related to players or player states, so you should probably name the class accordingly.

And I'm sure Alek would hate this class, because you're using a static class that has inner state (`playerList`). But I'll leave it to him to decide on what's best here.

File: QSB/QSB.csproj [hunk 2]
```diff
@@ -34,70 +34,79 @@
...
+      <HintPath>D:\EpicGames\OuterWilds\OuterWilds_Data\Managed\UnityEngine.UI.dll</HintPath>
     </Reference>
     <Reference Include="UnityEngine.UIModule, Version=0.0.0.0, Culture=neutral, processorArchitecture=MSIL">
       <SpecificVersion>False</SpecificVersion>
-      <HintPath>$(GameDir)\OuterWilds_Data\Managed\UnityEngine.UIModule.dll</HintPath>
+      <HintPath>D:\EpicGames\OuterWilds\OuterWilds_Data\Managed\UnityEngine.UIModule.dll</HintPath>
     </Reference>
     <Reference Include="UnityEngine.UNETModule">
-      <HintPath>$(GameDir)\OuterWilds_Data\Managed\UnityEngine.UNETModule.dll</HintPath>
+      <HintPath>E:\Steam\steamapps\common\Outer Wilds\OuterWilds_Data\Managed\UnityEngine.UNETModule.dll</HintPath>
```
Comment (Raicuparta) [inline QSB/QSB.csproj:109]: Can you revert this changes? Make sure the QSB.csproj never has any absolute paths from your system, it should always use $(GameDir) and $(OwmlDir).

You can commit your changes to QSB.csproj.user, that's fine.

File: QSB/TimeSync/PreserveTimeScale.cs [hunk 3]
```diff
@@ -8,8 +8,9 @@ public class PreserveTimeScale : NetworkBehaviour
     {
         private void Start()
         {
-            QSB.Helper.Menus.PauseMenu.GetButton("Button-EndCurrentLoop").Hide();
+            QSB.Helper.Menus.PauseMenu.GetTitleButton("Button-EndCurrentLoop").Hide(); // Remove the meditation button
```
Comment (Raicuparta) [inline QSB/TimeSync/PreserveTimeScale.cs:11]: I understand that this line might not have been clear before, so you added a comment, which is good. But for the future: you could have created a new method called `RemoveMeditationButton()` that just runs that one line. Now you have self-documenting code instead of code with documentation, which is always better :)

Check if it's possible to apply this whenever you want to add a comment to confusing code. 

There's definitely a lot of code in QSB that could use this treatment, especially for stuff inside lifecycle methods.

Lifecycle methods (Awake, Start, Update) are always gonna have non-descriptive names, because the name tells you when the method happens, not what the method does.

File: QSB/TimeSync/PreventShipDestruction.cs [hunk 4]
```diff
@@ -6,14 +6,17 @@
...
+    /// <summary>
+    /// Client-only-side component for stopping the ship from being damaged.
+    /// </summary>
     public class PreventShipDestruction : MonoBehaviour
     {
         private void Awake()
         {
-            QSB.Helper.HarmonyHelper.Transpile<ShipDetachableLeg>("Detach", typeof(Patch), "ReturnNull");
-            QSB.Helper.HarmonyHelper.Transpile<ShipDetachableModule>("Detach", typeof(Patch), "ReturnNull");
+            QSB.Helper.HarmonyHelper.Transpile<ShipDetachableLeg>("Detach", typeof(Patch), "ReturnNull"); // Stop legs from detaching
```
Comment (Raicuparta) [inline QSB/TimeSync/PreventShipDestruction.cs:16]: Another thing with having comments in the same line as the code, is that diffs become harder to read (longer line, and changes in the comment might look like changes in code at a first glance). 

Also means that if one commit makes a change to the comment and another one changes the code, it would count as a conflict.

Version-control-oriented programming 😬

File: QSB/TransformSync/PlayerCameraSync.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        public static PlayerCameraSync LocalInstance { get; private set; }
+
+        public override void OnStartLocalPlayer()
+        {
+            LocalInstance = this;
+        }
+
+        uint GetAttachedNetId()
+        {
+            return netId.Value - 2; // This is the 3rd transformsync in the "stack"
```
Comment (Raicuparta) [inline QSB/TransformSync/PlayerCameraSync.cs:21]: Interestiiing. Probably a good idea to leave a comment here explaining why this is needed to be done at all, since it is quite peculiar. Comments are more useful when they answer "why" instead of "what".

Plus I'd like to understand it out of curiosity.

File: QSB/TransformSync/PlayerHUDMarker.cs [hunk 6]
```diff
@@ -14,13 +14,12 @@ class PlayerHUDMarker : HUDDistanceMarker
...
-            _markerRadius = 0.3f;
+            _markerRadius = 2f;
 
             _markerTarget = new GameObject().transform;
             _markerTarget.parent = transform;
-            // I'm not really sure why this has to be 20 instead of 2 (the player height in units).
-            // But that's the only way it looks right.
-            _markerTarget.localPosition = Vector3.up * 20;
+
+            _markerTarget.localPosition = Vector3.zero;
```
Comment (Raicuparta) [inline QSB/TransformSync/PlayerHUDMarker.cs:22]: Oh. This makes it show above the player's head? Makes sense 😅 That's why it worked so weird.

File: QSB/Utility/PlayerToolsManager.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            var oldComponent = flashlightRoot.GetComponent<Flashlight>();
+            var component = flashlightRoot.AddComponent<QSBFlashlight>();
+            component._lights = oldComponent.GetValue<OWLight2[]>("_lights");
+            component._illuminationCheckLight = oldComponent.GetValue<OWLight2>("_illuminationCheckLight");
+            component._root = oldComponent.GetValue<Transform>("_root");
+            component._basePivot = oldComponent.GetValue<Transform>("_basePivot");
+            component._wobblePivot = oldComponent.GetValue<Transform>("_wobblePivot");
+            oldComponent.enabled = false;
+            flashlightRoot.transform.parent = _cameraBody;
+            flashlightRoot.transform.localPosition = new Vector3(0.7196316f, -0.2697681f, 0.3769455f);
```
Comment (Raicuparta) [inline QSB/Utility/PlayerToolsManager.cs:31]: maybe have this position as a readonly property up there

File: QSB/QSB.csproj (hunk 2 above)
Comment (misternebula) [reply QSB/QSB.csproj:109]: I.. did? i have no idea why that's happened

File: QSB/Animation/QSBFlashlight.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+		public Transform _wobblePivot;
+
+		private bool _flashlightOn;
+
+		private Vector3 _baseForward;
+
+		private Quaternion _baseRotation;
+	}
+}
```
[the remark this answers is not in the document]
Comment (misternebula) [reply QSB/Animation/QSBFlashlight.cs:72]: copied and edited yeah, i'll clean it up

File: QSB/Animation/QSBFlashlight.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        private Vector3 _baseForward;
+        private Quaternion _baseRotation;
+
+        private void Start()
+        {
+            _baseForward = _basePivot.forward;
+            _baseRotation = _basePivot.rotation;
+        }
+
+		public void TurnOn()
```
Comment (amazingalek) [inline QSB/Animation/QSBFlashlight.cs:26]: auto-format the document :-)
that goes for all documents, ofc

File: QSB/DebugLog.cs [hunk 10]
```diff
@@ -48,24 +54,23 @@ public static void Screen(params object[] logObjects)
...
 
-        public static void HUD(params object[] logObjects)
+        public static void ToHud(params object[] logObjects)
         {
             if (Locator.GetPlayerBody() == null)
             {
-                Console("Warning: tried to log to HUD but player is not ready.");
-                Console(logObjects);
+                //ToConsole("Warning: tried to log to HUD but player is not ready.");
+                //ToConsole("* " + JoinAll(logObjects));
```
Comment (amazingalek) [inline QSB/DebugLog.cs:62]: why commented?

File: QSB/Events/FullStateMessage.cs [hunk 11]
```diff
@@ -10,29 +10,18 @@ public class FullStateMessage : PlayerMessage
     {
         public override MessageType MessageType => MessageType.FullState;
 
-        public Dictionary<uint, string> PlayerNames;
+        public string PlayerName { get; set; }
```
Comment (amazingalek) [inline QSB/Events/FullStateMessage.cs:13]: no longer need the list?

File: QSB/Finder.cs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+using QSB.Events;
+using QSB.Utility;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+
+namespace QSB
+{
+    public static class PlayerRegistry
```
Comment (amazingalek) [inline QSB/Finder.cs:12]: the filename is something else

Comment (amazingalek) [reply QSB/Finder.cs:12]: This class has the potential to become a "god class" - a class that does all sorts of stuff. Not a huge issue right now, but is something to look out for.

File: QSB/QSB.csproj.user [hunk 13]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="Current" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <GameDir>C:\Program Files\Epic Games\OuterWilds</GameDir>
-    <OwmlDir>C:\Users\ricar\Repos\mod-installer\out\Outer Wilds Mod Manager-win32-x64\OWML</OwmlDir>
+    <GameDir>D:/EpicGames/OuterWilds</GameDir>
+    <OwmlDir>C:\Users\Henry\Downloads\OWModManager\OWML</OwmlDir>
```
Comment (amazingalek) [inline QSB/QSB.csproj.user:5]: Hello, Henry :D

File: QSB/TransformSync/PlayerCameraSync.cs (hunk 5 above)
Comment (amazingalek) [reply QSB/TransformSync/PlayerCameraSync.cs:21]: `GetComponent<TheMainNetworkThing>().netId.Value` instead?

File: QSB/TransformSync/SectorMessage.cs [hunk 14]
```diff
@@ -7,20 +7,19 @@ public class SectorMessage : QSBMessage
     {
         public override MessageType MessageType => MessageType.Sector;
 
-        public int SectorId;
+        public int SectorName;
```
Comment (amazingalek) [inline QSB/TransformSync/SectorMessage.cs:10]: how's a name an int?

File: QSB/TransformSync/ShipTransformSync.cs [hunk 15]
```diff
@@ -6,27 +6,40 @@ public class ShipTransformSync : TransformSync
...
             var shipModel = GetShipModel();
 
             var remoteTransform = new GameObject().transform;
 
             Instantiate(shipModel.Find("Module_Cockpit/Geo_Cockpit/Cockpit_Geometry/Cockpit_Exterior"), remoteTransform);
             Instantiate(shipModel.Find("Module_Cabin/Geo_Cabin/Cabin_Geometry/Cabin_Exterior"), remoteTransform);
+            Instantiate(shipModel.Find("Module_Cabin/Geo_Cabin/Cabin_Tech/Cabin_Tech_Exterior"), remoteTransform);
             Instantiate(shipModel.Find("Module_Supplies/Geo_Supplies/Supplies_Geometry/Supplies_Exterior"), remoteTransform);
             Instantiate(shipModel.Find("Module_Engine/Geo_Engine/Engine_Geometry/Engine_Exterior"), remoteTransform);
+
```
Comment (amazingalek) [inline QSB/TransformSync/ShipTransformSync.cs:43]: two newlines

File: QSB/TransformSync/TransformSync.cs [hunk 16]
```diff
@@ -78,18 +79,24 @@ private void Update()
                 return;
             }
 
-            var sectorTransform = SectorSync.Instance.GetSector(netId.Value);
+            // Get which sector should be used as a reference point
+            //var sectorTransform = SectorSync.Instance.GetSector(netId.Value);
+            //var sectorTransform = Finder.GetSector(netId.Value);
```
Comment (amazingalek) [inline QSB/TransformSync/TransformSync.cs:84]: commented code :P

File: QSB/Utility/FlagsHelper.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace QSB.Utility
+{
+    // Stolen from here : https://stackoverflow.com/questions/3261451/using-a-bitmask-in-c-sharp
+
+    public static class FlagsHelper
```
Comment (amazingalek) [inline QSB/Utility/FlagsHelper.cs:10]: What's this used for?

File: QSB/Utility/PlayerToolsManager.cs [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+            CreateFlashlight();
+        }
+
+        private static void CreateFlashlight()
+        {
+            var flashlightRoot = GameObject.Instantiate(GameObject.Find("FlashlightRoot"));
+            flashlightRoot.SetActive(false);
+            var oldComponent = flashlightRoot.GetComponent<Flashlight>();
+            var component = flashlightRoot.AddComponent<QSBFlashlight>();
```
Comment (amazingalek) [inline QSB/Utility/PlayerToolsManager.cs:24]: what's the point of the QSB version of flashlight?

File: QSB/manifest.json [hunk 19]
```diff
@@ -1,6 +1,6 @@
 {
   "filename": "QSB.dll",
-  "author": "Raicuparta",
+  "author": "Rai, Alek & Nebula",
```
Comment (amazingalek) [inline QSB/manifest.json:3]: <3

File: README.md [hunk 20]
```diff
@@ -68,12 +68,9 @@ If for some reason none of this is working, you might have to set everything man
 ## Authors
 
 - [AmazingAlek](https://github.com/amazingalek)
```
Comment (amazingalek) [inline README.md:70]: put me below :-)

**Comment by amazingalek** [comment]
This looks awesome! You've done amazing work! ❤️

File: QSB/DebugLog.cs (hunk 10 above)
Comment (misternebula) [reply QSB/DebugLog.cs:62]: it was annoying me while testing :P

File: QSB/Utility/FlagsHelper.cs (hunk 17 above)
Comment (misternebula) [reply QSB/Utility/FlagsHelper.cs:10]: using in playerregistry for changing states

File: QSB/TransformSync/PlayerCameraSync.cs (hunk 5 above)
Comment (misternebula) [reply QSB/TransformSync/PlayerCameraSync.cs:21]: the camera and player are different gameobjects, same with the ship

File: QSB/Utility/PlayerToolsManager.cs (hunk 18 above)
Comment (misternebula) [reply QSB/Utility/PlayerToolsManager.cs:24]: the normal flashlight can't be turned on/off without reflection (and would also control itself which is bad)

Comment (misternebula) [reply QSB/Utility/PlayerToolsManager.cs:24]: we want all the on/off logic to be performed on the local side, then send that info to the server

File: QSB/QSB.csproj.user (hunk 13 above)
Comment (misternebula) [reply QSB/QSB.csproj.user:5]: 👋

File: QSB/Finder.cs (hunk 12 above)
Comment (misternebula) [reply QSB/Finder.cs]: fixed file name - keeping unresolved to remind me not to over complicate it :P

File: QSB/Utility/PlayerToolsManager.cs (hunk 18 above)
Comment (amazingalek) [reply QSB/Utility/PlayerToolsManager.cs:24]: wouldn't it be simpler to use some reflection and maybe empty a method or two to disable the automatic stuff?

File: QSB/TransformSync/PlayerCameraSync.cs (hunk 5 above)
Comment (amazingalek) [reply QSB/TransformSync/PlayerCameraSync.cs:21]: We're considering the netId of the *main network thing* of a player to the the *player ID*, right? If so, maybe PlayerRegistry could keep track of those, so we could just say PlayerRegistry.MyId or similar?

Comment (amazingalek) [reply QSB/TransformSync/PlayerCameraSync.cs:21]: Yeah, that doesn't solve this issue...
Anyway, the current solution seems fragile, will break if component order changes, right?

File: QSB/Utility/PlayerToolsManager.cs (hunk 18 above)
Comment (Raicuparta) [reply QSB/Utility/PlayerToolsManager.cs:24]: I think it makes sense to have QSBFlashlight, but maybe it should be a much simpler flashlight than the game's.

Maybe we shouldn't just instantiate the entire flashlight object, and instead have only an object with one light source and nothing more. Then you can just enable / disable that object for toggling the flashlight.

The current in-game flashlight contains multiple light sources (maybe just 2, but at least 2) and casts shadows, so having lots of players with lots of flashlights might cause performance issues.

Would be better to save this for a future PR, though.

And @amazingalek patching methods would be difficult because you would also affect the local player's flashlight.
